### ISSUE TITLE
Compiled `--erased-cubical` program produces non-deterministic value

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#7466]
Lifted from #7442.
Consider these two modules:
A `--cubical` module:
```agda
{-# OPTIONS --erasure --cubical #-}

module CubicalUnit where

record One : Set where
  constructor one
```
Imported by a `--erased-cubical` module:
```agda
{-# OPTIONS --erasure --erased-cubical #-}

module TestCubicalUnit where

open import Common.Prelude
open import CubicalUnit

f : One → Nat
f one = 42

postulate t : One

main = printNat (f t)
```
This compiles fine with the GHC backend and prints some number non-deterministically, e.g. `4300275432`.

The expected result `42` can be produced by e.g. one of these changes:
1. Inline the definition of `record One` rather than importing it.
2. Change the type of `f` to `@0 One -> Nat`.

The source of the problem seems to be that clauses like `f one = 42` matching on an erased constructor `one` are simply removed by the compiler, i.e., replaced by `f = erased`.  Via the unsafe coercions in the compiled code we do not crash, but coerce some pointers(?) to an integer we can print.

I find it weird that `f` is allowed to match on an erased constructor without its domain being marked as erased.

ATTN: @nad.

**Comment by andreasabel** [comment]
On a related note, `f t` evaluates to `42` in interactive mode.  This indicates that `--erased-cubical` does treat `one` as an eta constructor even though it is erased.  
However, commit 5c4ab96071a7118a8e0b379ecd7eafeee2f2dcfb by @nad, fix of #6287 suggests that erased constructors should maybe (at least sometimes) not have eta.  It is not clear to me in which situations eta is desired and in which not.  A clear theory seems to be lacking yet and maybe the fix of #6287 was not systematic.
- #6287

**Comment by nad** [comment]
Isn't the problem here that you postulate that there is a non-erased inhabitant of a type that should only contain erased inhabitants?

> I find it weird that `f` is allowed to match on an erased constructor without its domain being marked as erased.

See my [other answer](https://github.com/agda/agda/issues/7442#issuecomment-2326528352).

> On a related note, `f t` evaluates to `42` in interactive mode. This indicates that `--erased-cubical` does treat `one` as an eta constructor even though it is erased.

*Run-time* irrelevance annotations (like `@0`) should not affect how things compute at compile-time.

> However, commit [5c4ab96](https://github.com/agda/agda/commit/5c4ab96071a7118a8e0b379ecd7eafeee2f2dcfb) by @nad, fix of #6287 suggests that erased constructors should maybe (at least sometimes) not have eta. It is not clear to me in which situations eta is desired and in which not. A clear theory seems to be lacking yet and maybe the fix of #6287 was not systematic.

See my [other answer](https://github.com/agda/agda/issues/7442#issuecomment-2326528352).

**Comment by andreasabel** [comment]
@UlfNorell argued the same thing would happen with `data One : Set where one : One`, so it isn't anything specific to records.  (Except that then `f t` would not evaluate to `42` at compile time, but be stuck.)

Option `--safe` would prevent my counterexample.   A problem is that `--safe` is practically unusable with compilation because it rules out usage of the FFI.  But this is another issue...
